tidy3d/components/data/sim_data.py

```diff
@@ -953,52 +980,256 @@ def source_spectrum_fn(freqs):
             wavelength_original = grid_spec_original.wavelength_from_sources(sim_original.sources)
             grid_spec_adj = grid_spec_original.updated_copy(wavelength=wavelength_original)
             sim_adj_update_dict["grid_spec"] = grid_spec_adj
 
-        return sim_original.updated_copy(**sim_adj_update_dict)
+        return sim_original.updated_copy(**sim_adj_update_dict), adjoint_source_info
 
-    def make_adjoint_sources(self, data_vjp_paths: set[tuple]) -> list[Source]:
+    def make_adjoint_sources(self, data_vjp_paths: set[tuple]) -> dict[str, Source]:
         """Generate all of the non-zero sources for the adjoint simulation given the VJP data."""
 
-        # TODO: determine if we can do multi-frequency sources
-
         # map of index into 'self.data' to the list of datasets we need adjoint sources for
         adj_src_map = defaultdict(list)
         for _, index, dataset_name in data_vjp_paths:
             adj_src_map[index].append(dataset_name)
 
-        # gather a list of adjoint sources for every monitor data in the VJP that needs one
-        sources_adj_all = []
+        # gather a dict of adjoint sources for every monitor data in the VJP that needs one
+        sources_adj_all = defaultdict(list)
         for data_index, dataset_names in adj_src_map.items():
             mnt_data = self.data[data_index]
-            sources_adj = mnt_data.make_adjoint_sources(dataset_names=dataset_names)
-            sources_adj_all += sources_adj
+            sources_adj = mnt_data.make_adjoint_sources(
+                dataset_names=dataset_names, fwidth=self.fwidth_adj
+            )
+            sources_adj_all[mnt_data.monitor.name] = sources_adj
 
         return sources_adj_all
 
+    @property
+    def fwidth_adj(self) -> float:
+        # fwidth of forward pass, try as default for adjoint
+        normalize_index_fwd = self.simulation.normalize_index or 0
+        return self.simulation.sources[normalize_index_fwd].source_time.fwidth
+
+    def process_adjoint_sources(self, adj_srcs: list[Source]) -> AdjointSourceInfo:
+        # tuple[list[Source], Union[float, xr.DataArray], bool]:
+        """Compute list of final sources along with a post run normalization for adj fields."""
+
+        # dictionary mapping unique spatial dependence of each Source to list of time-dependencies
+        json_to_sources = defaultdict(None)
+        spatial_to_src_times = defaultdict(list)
+        for src in adj_srcs:
+            src_spatial_json = src.json(exclude={"source_time"})
+            json_to_sources[src_spatial_json] = src
+            spatial_to_src_times[src_spatial_json].append(src.source_time)
+
+        num_ports = len(spatial_to_src_times)
+        num_unique_freqs = len({src.source_time.freq0 for src in adj_srcs})
+
+        # next, figure out which treatment / normalization to apply
+        if num_unique_freqs == 1:
+            log.info("adjoint source creation: one unique frequency, no normalization")
```

Comment:
```Use log.info to log general information about the application's execution, such as key events and major steps, which are useful for end users or system administrators. Use log.debug for detailed diagnostic information useful for developers, such as internal states and variable values during execution, to troubleshoot specific issues.
```
from chat GPT, to me it seems like this falls under info? because it's more generally something that the user might want to see and also it basically just tells us what major steps are happening internally?